### PR TITLE
Reworked message routing into IMqttBridge

### PR DESCRIPTION
**Pull request by nayato**
Motivation:
previous topic name routing did nothing really and allowed for topic name conversion in a single fixed way.

Modifications:
- moved topic name conversion from MqttIotHubAdapter to IotHubClient
- folded messaging factory into messaging client
- renamed MqttIotHubAdapter -> MqttAdapter
- introduced IMqttCommunicator that allows for election of handling messaging client and sourcing c2d messages from multiple messaging clients
- switched to push model for c2d messages
- moving throttling of c2d messages into IotHubClient (using PrefetchCount on Device Client to limit message passing)
- extra: explicitly calling out a case when retransmission is on and there's more than one c2d pending message allowed (not supported)

Result:
protocol gateway core is ready for flexible traffic routing (d2c) and aggregation (c2d), better separation of concerns between messaging client and MqttAdapter

### CODE REVIEW

File: host/ProtocolGateway.Host.Cloud.Service/mqttTopicRouting.config.user [hunk 1]
```diff
@@ -1,8 +0,0 @@
-<mqttTopicRouting>
```
Comment (mtuchkov) [inline host/ProtocolGateway.Host.Cloud.Service/mqttTopicRouting.config.user]: BTW, This is a breaking change, is it OK?

File: host/ProtocolGateway.Host.Common/Bootstrapper.cs [hunk 2]
```diff
@@ -53,16 +55,14 @@ public Bootstrapper(ISettingsProvider settingsProvider, ISessionStatePersistence
 
             this.settingsProvider = settingsProvider;
             this.settings = new Settings(this.settingsProvider);
+            this.iotHubClientSettings = new IotHubClientSettings(this.settingsProvider);
             this.sessionStateManager = sessionStateManager;
             this.qos2StateProvider = qos2StateProvider;
             this.authProvider = new SasTokenDeviceIdentityProvider();
-            this.iotHubMessageRouter = new ConfigurableMessageRouter();
+            this.topicNameConverter = new ConfigurableMessageAddressConverter();
```
Comment (mtuchkov) [inline host/ProtocolGateway.Host.Common/Bootstrapper.cs]: Do we need to have 'Configurable' prefix?

File: host/ProtocolGateway.Host.Common/ProtocolGateway.Host.Common.csproj [hunk 3]
```diff
@@ -104,6 +96,16 @@
       <SubType>Designer</SubType>
     </None>
   </ItemGroup>
+  <ItemGroup>
```
Comment (mtuchkov) [inline host/ProtocolGateway.Host.Common/ProtocolGateway.Host.Common.csproj]: Have you convince Konstantin?

File: host/ProtocolGateway.Host.Cloud.Service/mqttTopicRouting.config.user (hunk 1 above)
Comment (nayato) [inline host/ProtocolGateway.Host.Cloud.Service/mqttTopicRouting.config.user]: will make it v1.1.

File: src/ProtocolGateway.Core/Messaging/SingleClientCommunicator.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.Azure.Devices.ProtocolGateway.Messaging
+{
+    using System;
+    using System.Threading.Tasks;
+
+    public sealed class SingleClientCommunicator : IMqttCommunicator, IMessagingChannel<IMessage>
```
Comment (mtuchkov) [inline src/ProtocolGateway.Core/Messaging/SingleClientCommunicator.cs]: I think this is a Bridge

File: host/ProtocolGateway.Host.Common/ProtocolGateway.Host.Common.csproj (hunk 3 above)
Comment (nayato) [inline host/ProtocolGateway.Host.Common/ProtocolGateway.Host.Common.csproj]: will revert

File: host/ProtocolGateway.Host.Common/Bootstrapper.cs (hunk 2 above)
Comment (nayato) [inline host/ProtocolGateway.Host.Common/Bootstrapper.cs]: erm.. without it "I" is the only thing separating it from interface. Should have at least some specialization. Other ideas?

File: src/ProtocolGateway.Core/Mqtt/AckPendingMessageState.cs [hunk 5]
```diff
@@ -10,56 +10,40 @@ namespace Microsoft.Azure.Devices.ProtocolGateway.Mqtt
...
-        public AckPendingMessageState(ulong sequenceNumber, int packetId, QualityOfService qualityOfService, string lockToken)
-        {
-            this.SequenceNumber = sequenceNumber;
-            this.PacketId = packetId;
-            this.QualityOfService = qualityOfService;
-            this.LockToken = lockToken;
+            this.SequenceNumber = messageWithFeedback.Message.SequenceNumber;
+            this.PacketId = packet.PacketId;
+            this.QualityOfService = packet.QualityOfService;
+            this.TopicName = preserveTopicName ? messageWithFeedback.Message.Address : null;
```
Comment (mtuchkov) [inline src/ProtocolGateway.Core/Mqtt/AckPendingMessageState.cs]: How does it work?

Comment (nayato) [inline src/ProtocolGateway.Core/Mqtt/AckPendingMessageState.cs]: it isn't needed anymore actually, will update.

File: src/ProtocolGateway.IotHubClient/IotHubClient.cs [hunk 6]
```diff
@@ -57,82 +65,148 @@ public IotHubClient(DeviceClient deviceClient)
...
+        public IMessage CreateMessage(string address, Stream payload) => new IotHubClientMessage(address, new Message(payload));
+
+        public void BindMessagingChannel(IMessagingChannel<IMessage> channel)
+        {
+            Contract.Requires(channel != null);
+
+            Contract.Assert(this.messagingChannel == null);
+
+            this.messagingChannel = channel;
+            this.Receive();
```
Comment (mtuchkov) [inline src/ProtocolGateway.IotHubClient/IotHubClient.cs]: Looks like a side-effect

File: src/ProtocolGateway.IotHubClient/IotHubClient.cs [hunk 7]
```diff
@@ -57,82 +65,148 @@ public IotHubClient(DeviceClient deviceClient)
...
         }
 
         public async Task SendAsync(IMessage message)
         {
             try
             {
-                await this.deviceClient.SendEventAsync(((DeviceClientMessage)message).ToMessage());
+                string address = message.Address;
+                if (this.messageAddressConverter.TryParseAddressIntoMessageProperties(address, message))
+                {
```
Comment (mtuchkov) [inline src/ProtocolGateway.IotHubClient/IotHubClient.cs]: Should we do something here?

File: src/ProtocolGateway.IotHubClient/IotHubClient.cs [hunk 8]
```diff
@@ -57,82 +65,148 @@ public IotHubClient(DeviceClient deviceClient)
...
+                    if (CommonEventSource.Log.IsWarningEnabled)
+                    {
+                        CommonEventSource.Log.Warning("Topic name could not be matched against any of the configured routes. Falling back to default telemetry settings.", address);
+                    }
+                    message.Properties[this.settings.ServicePropertyPrefix + MessagePropertyNames.Unmatched] = bool.TrueString;
+                    message.Properties[this.settings.ServicePropertyPrefix + MessagePropertyNames.Subject] = address;
+                }
+
+                string messageDeviceId;
+                if (message.Properties.TryGetValue(TemplateParameters.DeviceIdTemplateParam, out messageDeviceId))
```
Comment (mtuchkov) [inline src/ProtocolGateway.IotHubClient/IotHubClient.cs]: Is this correct?

Comment (nayato) [inline src/ProtocolGateway.IotHubClient/IotHubClient.cs]: just moved inside if clause per previous comment.

File: src/ProtocolGateway.IotHubClient/IotHubClient.cs (hunk 7 above)
Comment (nayato) [inline src/ProtocolGateway.IotHubClient/IotHubClient.cs]: now we do.

**Comment by mtuchkov** [comment]
:shipit:

**Pull request merged**